### PR TITLE
Cache inflight gets and share cache between sessions

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request caches inflight gets, rather than only cache the finished request, and shares the cache between all sessions.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -587,12 +590,13 @@ module.exports = class Hypercore extends EventEmitter {
...
-    if (c) return c
-    const fork = this.core.tree.fork
-    const b = await this._get(index, opts)
-    if (this.cache && fork === this.core.tree.fork && b) this.cache.set(index, b)
-    return b
+    let block = this.cache && this.cache.get(index)
+    if (block) return block
+
+    block = this._get(index, opts)
+    if (this.cache) this.cache.set(index, block)
```
[the remark this answers is not in the document]
Comment (mafintosh) [reply index.js:597]: ah we can't do that here, cause that caches the replication promise which gets cancelled on session close. the good news is that the replication promises are already shared internally and do no io, so i __think__ you just need to move this line down into _get where it loads from the block store

[the remark this answers is not in the document]
Comment (kasperisager) [reply index.js:597]: Yep, so two gets for the same block will yield the exact same promise and not issue another disk read or replicator request. As the method itself is async, the promise chain is flattened.

[the remark this answers is not in the document]
Comment (kasperisager) [reply index.js:597]: Perfect, I'll move stuff around 👍

[the remark this answers is not in the document]
Comment (davidmarkclements) [reply index.js:597]: I don't *think* there's any issue with caching promises *as long as* they never throw unhandled promise rejections

[the remark this answers is not in the document]
Comment (kasperisager) [reply index.js:597]: They're always passed on to the caller as well for error handling.

**Comment by kasperisager** [comment]
I still have some corner cases to sort out.

**Pull request merged**